### ISSUE TITLE
git:// urls are no longer supported on GitHub - Breaks build

### ISSUE THREAD

**Issue by mw9** [ralph-irving/squeezeos#15]
`git://` urls are no longer supported on GitHub. This is breaking the build (at least on my setup).

Refer section _No more unauthenticated Git_ in this blog note _Improving Git protocol security on GitHub_:
https://github.blog/2021-09-01-improving-git-protocol-security-github/

> On the Git protocol side, unencrypted git:// offers no integrity or authentication, making it subject to tampering. We expect very few people are still using this protocol, especially given that you can’t push (it’s read-only on GitHub). We’ll be disabling support for this protocol.

Directly impacted is: `poky/meta-squeezeos/conf/distro/squeezeos.conf`, which contains the following `git://` urls:
 ```
# SqueezeOS/SqueezePlay SCM
SQUEEZEOS_SVN ?= "git://github.com/ralph-irving/squeezeos;branch=public/${DISTRO_VERSION};repopath=src;protocol=git"
SQUEEZEPLAY_SCM ?= "git://github.com/ralph-irving/squeezeos-squeezeplay.git;branch=public/${DISTRO_VERSION};repopath=src;protocol=git"
 
 ##### Force 7.7.3
 ##### SQUEEZECENTER_SCM ?= "git://github.com/Logitech/slimserver.git;branch=public/${DISTRO_VERSION};protocol=git"
SQUEEZECENTER_SCM ?= "git://github.com/ralph-irving/slimserver.git;tag=${DISTRO_RELEASE};protocol=git"
```
This can be easily worked around by changing each `protocol=git` to `protocol=https`. The `git://` prefix should stay, that just tells Bitbake that we have a git repository, so it is still needed.

There may be other recipes, etc., impacted but I have not needed to find them...

**BUT**:

The version of git (v1.5.2.3) in use is too old to handle http(s) transport on GitHub. It uses 'Dumb HTTP Transport', and GitHub banished that about ten years ago. Refer this blog note _Git Dumb HTTP Transport to be turned off in 90 days_:
https://github.blog/2011-03-09-git-dumb-http-transport-to-be-turned-off-in-90-days/

> Git v1.6.6 (released January, 2010) added a “smart” HTTP transport that is superior to the old transport in every way

Well, I tried updating the git package to v1.7.2.1, which I found in the Poky/Yocto repository. But that doesn't work, because it seems to require an outdated version of openssl (perhaps 1.02). I haven't pursued updating git further, but something needs to be done.

Pro-tem, the something I am doing is to use the the version of git (2.25.0) that you kindly provided a while ago, installed in /usr/local/bin. It seems to be working just fine.

To activate it, though, needs a package to be installed. I have drafted one.

**Comment by mw9** [comment]
Here is a draft of the package that I have installed as `poky/meta-squeezeos/packages/git/local-git.bb`. It seems to work fine, when enabled as instructed.

As far as I can tell, nothing awkward seems to be taking place with the build by using this more modern git. But may require some testing.

```
DESCRIPTION = "The git revision control system used by the Linux kernel developers"
SECTION = "console/utils"
LICENSE = "GPL"

PROVIDES = "git-native"

# This recipe installs a link to the locally installed git in /usr/local/bin for
# use in the build process. This removes the requirement to build git from source.

# Enable it by setting PREFERRED_PROVIDER_git-native = "local-git" in the
# configuration file poky/build/conf/site.conf.

# Disable it by setting PREFERRED_PROVIDER_git-native = "git-native".

# Bitbake creates a dependency on package git-native for all packages that use
# the git protocol. So, to use a pre-installed git, we must have a package in
# place that installs a suitable link in the staging hierarchy. (Which is where
# build host programs are run from).

inherit native

do_stage () {
	ln -sf /usr/local/bin/git ${STAGING_BINDIR}/git
}

```

**Comment by ralph-irving** [comment]
Thanks for taking the time to put this together.
I haven't built the firmware in a while.
I'll add your recipe and search the controller and touch specific packages for similiar.

**Comment by ralph-irving** [comment]
I have updated the git recipe to 2.25.5, openssl to 1.1.1o and curl to 7.52.1 on my built system and successfully rebuilt all 3 firmware types.  So we have a longer term fix.  I'd like to track down why poky/bitbake/lib/bb/fetch/wget.py picks up the /usr/bin/wget binary never searching /usr/local/bin first for the "working" one before pushing the changes.

**Comment by mw9** [comment]
Well done !

I think I can answer your `wget` question. 

`bitbake` seems to empty the environment before it runs.
Refer: https://github.com/Logitech/squeezeos/blob/585fb6501704f0ca93e2bde662ccf4f8c4c9af10/poky/bitbake/lib/bb/cooker.py#L588

It calls `wget` thusly:  `/usr/bin/env wget ..`.
Refer: https://github.com/Logitech/squeezeos/blob/585fb6501704f0ca93e2bde662ccf4f8c4c9af10/poky/meta/conf/bitbake.conf#L480

`env` appears to use `execvp` to do the deed. `execvp` will search `PATH` if it is defined, otherwise it uses a system default. That system default does not include `/usr/local/bin` on my Ubuntu 10.04, nor, I think, on my macOS.

So, I think it is the clearance of the PATH environment variable by `bitbake` that is the root cause.

Here is what I have in my `poky/build/conf/site.conf` to cover the point:
```
### Overides to meta/conf/bitbake.conf
                
            # Pick up our local wget. bitbake does not provide system PATH when attempting
            # to execute wget, so our local version is never found.

            FETCHCMD_wget = "/usr/local/bin/wget -t 5"
            FETCHCOMMAND_wget = "/usr/local/bin/wget -t 5 --passive-ftp -P ${DL_DIR} '${URI}'"
            CHECKCOMMAND_wget = "/usr/local/bin/wget --spider -t 5 --passive-ftp -P ${DL_DIR} '${URI}'"
            RESUMECOMMAND_wget = "/usr/local/bin/wget -c -t 5 --passive-ftp -P ${DL_DIR} '${URI}'"
   
``` 

I just checked some, probably outdated, manual pages on line. They suggest that I am on the right track.

https://man7.org/linux/man-pages/man3/exec.3.html
https://man7.org/linux/man-pages/man3/confstr.3.html

**Comment by mw9** [comment]
Thanks for the new radio firmware, I'll give that a go and report back.

One thing, though. I noticed your commit [Update timezone recipes to 2021e](https://github.com/ralph-irving/squeezeos/commit/4b771991dc8dfdcdbe4c1cff9b62a74cbd983cfd)  a few days ago.

I don't want to be a killjoy, but SqueezeOS doesn't use this package. Not for the Radio, anyway.

I think timezone data, etc, may be provided by the Code Sourcery toolchain, but I don't know how that works.

Or perhaps I'm missing something.

**Comment by ralph-irving** [comment]
Firstly, thank you for the detailed info on the wget issue.  I had renamed /usr/bin/wget and linked it to the one in local many years ago and quickly moved on.  I noticed it recently and changed it back only to redicovery why I changed it in the first place...to lazy at the time to figure out the root cause.
I was thinking it would be good to handle the wget issue in some way.  Like building wget-native and dependencies before starting the firmware build.  It would mean hosting the required source files on a webserver that supports http: or ftp: without the automatic https: redirect.  But I don't think it's worth the effort.  You've so kindly documented the issue and a workaround that should be enough.  I think that we should include the above commented out in the site.conf.sample file.

I'm pretty sure that the firmware files in /usr/share/zoneinfo are populated from the tzdata-[region] ipk packages.
You should have this folder in tmp-baby/work/armv5te-none-linux-gnueabi/tzdata-2021e-r0 after a firmware build.
Now if any part of the firmware actually uses them other than setting a symbolic link from etc when you set the timezone in the squeezeplay settings is another question.

**Comment by mw9** [comment]
**wget**
> I think that we should include the above commented out in the site.conf.sample file.

I agree. I think that is the most economical approach.

**tzdata**

I'm certain that `/usr/share/zoneinfo` is **NOT** populated from the tzdata-[region] ipk packages.

My recent clean builds (after `rm tmp-baby; rm sources`) simply do not download this package. There is no `tmp-baby/work/armv5te-none-linux-gnueabi/tzdata-2021e-r0` on my build machine.

I am seeing no changed files in the `zoneinfo` directory within the new firmware file you have just prepared. Unless I have goofed somehow.

I do, however, clearly see `tmp-baby/work/armv5te-none-linux-gnueabi/external-csl-toolchain-2010q1-202-r2/image/usr/share/zoneinfo`  on my build machine.
 
I don't have any good knowledge of how the packaging works, but I'd be mightily surprised if `zoneinfo` is coming from anywhere other than the Code Sourcery toolchain.

> Now if any part of the firmware actually uses them other than setting a symbolic link from etc when you set the timezone in the squeezeplay settings is another question.

I haven't sought to answer that question, and share your thinking. :smiley:
From memory SqueezePlay derives its sense of time from LMS. Or perhaps that is only a partial truth.

**Comment by ralph-irving** [comment]
Great!  I'll add the info to site.conf.sample.

Well, looking at it for the third time, I have screwed something up as I delete the tmp-* folders before a full rebuild and the tzdata-2021e-r0 directory is there afterward however it's NOT populated in tmp-fab4 or tmp-jive.  Something I will need to track down.

I'll remove the item from the changelog and leave the newer timezone recipe in place.  It shouldn't impact anything.

**Comment by ralph-irving** [comment]
I've pushed the changes required to update git.  Hopefully I haven't broken anything else.

**Comment by mw9** [comment]
> I've pushed the changes required to update git. Hopefully I haven't broken anything else.

- **Checksum error for jpegsrc.v8b.tar.gz**

I'm not seeing the same checksums for `jpegsrc.v8b.tar.gz` that you loaded up in commit https://github.com/ralph-irving/squeezeos/commit/8e775c7aaa8229dd20f1210c5c4bbbda4cbbeea5:

You've entered:
```
[http://www.ijg.org/files/jpegsrc.v8b.tar.gz]
 md5=e022acbc5b36cd2cb70785f5b575661e
 sha256=36e6208edec591bae8f2fc370ea4f991447badb6377a125c211ffa7b503174a7
```
But I'm getting:
```
wget http://www.ijg.org/files/jpegsrc.v8b.tar.gz
md5sum jpegsrc.v8b.tar.gz
61ea7b0b1489b6a0315faedb081f5b45  jpegsrc.v8b.tar.gz
shasum -a 256 jpegsrc.v8b.tar.gz
64be2429346b1aa84a014fc47264661ca2e6c786aea25afecb363d5973b4e6a9  jpegsrc.v8b.tar.gz
```
I need to change your checksums for the build to succeed.

- **Build failure git v2.25.5**

Building git v2.25.5 failed on a clean build system, due to a compile time dependency on `gettext-native`.

```
|     SUBDIR templates
|     MSGFMT po/build/locale/bg/LC_MESSAGES/git.mo
| /bin/sh: msgfmt: command not found
| make: *** [po/build/locale/bg/LC_MESSAGES/git.mo] Error 127
| FATAL: oe_runmake failed
NOTE: Task failed: /home/martin/squeezeos/poky/build/tmp-baby/work/i686-linux/git-native-2.25.5-r0/temp/log.do_compile.13525
NOTE: package git-native-2.25.5-r0: task do_compile: failed
```
`msgfmt` is provided by `gettext`.

We need `gettext-native` to be available before compiling git.

I would propose amending `git.inc` from:
`DEPENDS = "perl-native openssl curl zlib expat"`
to:
`DEPENDS = "perl-native openssl curl zlib expat gettext-native"`.

That cured it here.

**Comment by ralph-irving** [comment]
Thanks for testing the changes.

I have gettext installed from the Ubuntu OS packages, so that explains why I didn't see that failure.  It must have been installed as a dependency at some point.

```
$ dpkg -l | grep gettext
ii  gettext                              0.17-8ubuntu3                                   GNU Internationalization utilities
ii  gettext-base                         0.17-8ubuntu3                                   GNU Internationalization utilities for the b
ii  liblocale-gettext-perl               1.05-6                                          Using libc functions for internationalizatio

```
I did the checksums on the jpeg source file that was already in my sources folder and has likely been there since it was first downloaded. So that's similiar to what you described earlier where some other checksums had changed.
```
ralphy@poky:~/source/squeezeos/poky/sources$ ls -l jpegsrc.v8b.tar.gz
-rw-r--r-- 1 ralphy ralphy 965125 2010-05-16 04:23 jpegsrc.v8b.tar.gz
ralphy@poky:~/source/squeezeos/poky/sources$ md5sum jpegsrc.v8b.tar.gz && sha256sum jpegsrc.v8b.tar.gz
e022acbc5b36cd2cb70785f5b575661e  jpegsrc.v8b.tar.gz
36e6208edec591bae8f2fc370ea4f991447badb6377a125c211ffa7b503174a7  jpegsrc.v8b.tar.gz
```
I downloaded the jpeg source again and get the same checksum as you and the file has been changed.
```
ralphy@poky:~/source/squeezeos/poky/sources$ ls -l jpegsrc.v8b.tar.gz*
-rw-r--r-- 1 ralphy ralphy 935594 2022-01-04 05:28 jpegsrc.v8b.tar.gz
-rw-r--r-- 1 ralphy ralphy 965125 2010-05-16 04:23 jpegsrc.v8b.tar.gz.orig
```

**Comment by ralph-irving** [comment]
I just tried to rebuild git with gettext-native added and it failed to download the gnutls-2.4.2.tar.bz2 sources.
```NOTE: package gnutls-2.4.2-r0: task do_fetch: started
NOTE: fetch ftp://ftp.gnutls.org/pub/gnutls/gnutls-2.4.2.tar.bz2
--2022-05-15 07:36:26--  ftp://ftp.gnutls.org/pub/gnutls/gnutls-2.4.2.tar.bz2
           => '/home/ralphy/source/squeezeos/poky/sources/gnutls-2.4.2.tar.bz2'
Resolving ftp.gnutls.org... 217.69.76.55
Connecting to ftp.gnutls.org|217.69.76.55|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /pub/gnutls ... 
No such directory 'pub/gnutls'.

NOTE: fetch ftp://ftp.gnupg.org/gcrypt/gnutls/gnutls-2.4.2.tar.bz2
--2022-05-15 07:36:27--  ftp://ftp.gnupg.org/gcrypt/gnutls/gnutls-2.4.2.tar.bz2
           => '/home/ralphy/source/squeezeos/poky/sources/gnutls-2.4.2.tar.bz2'
Resolving ftp.gnupg.org... 217.69.76.55
Connecting to ftp.gnupg.org|217.69.76.55|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /gcrypt/gnutls ... done.
==> SIZE gnutls-2.4.2.tar.bz2 ... done.

==> PASV ... done.    ==> RETR gnutls-2.4.2.tar.bz2 ... 
No such file 'gnutls-2.4.2.tar.bz2'.

NOTE: fetch http://www.mirrors.wiretapped.net/security/network-security/gnutls/gnutls-2.4.2.tar.bz2
--2022-05-15 07:36:28--  http://www.mirrors.wiretapped.net/security/network-security/gnutls/gnutls-2.4.2.tar.bz2
Resolving www.mirrors.wiretapped.net... failed: Name or service not known.
wget: unable to resolve host address 'www.mirrors.wiretapped.net'
NOTE: fetch ftp://ftp.mirrors.wiretapped.net/pub/security/network-security/gnutls/gnutls-2.4.2.tar.bz2
--2022-05-15 07:36:28--  ftp://ftp.mirrors.wiretapped.net/pub/security/network-security/gnutls/gnutls-2.4.2.tar.bz2
           => '/home/ralphy/source/squeezeos/poky/sources/gnutls-2.4.2.tar.bz2'
Resolving ftp.mirrors.wiretapped.net... failed: Name or service not known.
wget: unable to resolve host address 'ftp.mirrors.wiretapped.net'
NOTE: fetch http://josefsson.org/gnutls/releases/gnutls-2.4.2.tar.bz2
--2022-05-15 07:36:28--  http://josefsson.org/gnutls/releases/gnutls-2.4.2.tar.bz2
Resolving josefsson.org... 178.174.241.102, 2001:9b1:8633::102
Connecting to josefsson.org|178.174.241.102|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://josefsson.org/gnutls/releases/gnutls-2.4.2.tar.bz2 [following]
--2022-05-15 07:36:29--  https://josefsson.org/gnutls/releases/gnutls-2.4.2.tar.bz2
Connecting to josefsson.org|178.174.241.102|:443... connected.
WARNING: cannot verify josefsson.org's certificate, issued by 'CN=R3,O=Let\'s Encrypt,C=US':
  Unable to locally verify the issuer's authority.
HTTP request sent, awaiting response... 404 Not Found
2022-05-15 07:36:29 ERROR 404: Not Found.

NOTE: fetch http://downloads.slimdevices.com/poky-cache/gnutls-2.4.2.tar.bz2
URL transformed to HTTPS due to an HSTS policy
--2022-05-15 07:36:29--  https://downloads.slimdevices.com/poky-cache/gnutls-2.4.2.tar.bz2
Resolving downloads.slimdevices.com... 13.226.142.111, 13.226.142.112, 13.226.142.105, ...
Connecting to downloads.slimdevices.com|13.226.142.111|:443... connected.
WARNING: cannot verify downloads.slimdevices.com's certificate, issued by 'CN=Amazon,OU=Server CA 1B,O=Amazon,C=US':
  Unable to locally verify the issuer's authority.
HTTP request sent, awaiting response... 404 Not Found
2022-05-15 07:36:30 ERROR 404: Not Found.

NOTE: Task failed: Fetch failed: ftp://ftp.gnutls.org/pub/gnutls/gnutls-2.4.2.tar.bz2
NOTE: package gnutls-2.4.2-r0: task do_fetch: failed
ERROR: TaskFailed event exception, aborting
NOTE: package gnutls-2.4.2: failed
ERROR: Build of /home/ralphy/source/squeezeos/poky/meta/packages/gnutls/gnutls_2.4.2.bb do_fetch failed
ERROR: Task 331 (/home/ralphy/source/squeezeos/poky/meta/packages/gnutls/gnutls_2.4.2.bb, do_fetch) failed
```
Did you have that issue?

**Comment by mw9** [comment]
> I downloaded the jpeg source again and get the same checksum as you and the file has been changed.

Do the contents stack up ? I found no changes there, so I guess it might have been a  change in packaging rather than content.

**Comment by mw9** [comment]
> I just tried to rebuild git with gettext-native added and it failed to download the gnutls-2.4.2.tar.bz2 sources.
> ...
> Did you have that issue?

No. There's no sign of my build(s) trying to download  gnutls, let alone build it.

Which is puzzling, because both curl_7.18.0.bb and curl_7.52.1.bb depend on it.

Confused...

**Comment by ralph-irving** [comment]
> > I downloaded the jpeg source again and get the same checksum as you and the file has been changed.
> 
> Do the contents stack up ? I found no changes there, so I guess it might have been a change in packaging rather than content.

The newer jpeg sources have changes to the Visual Studio Express 2010 project files.  All the jpeg source code is the same.

**Comment by mw9** [comment]
> No. There's no sign of my build(s) trying to download gnutls, let alone build it.
> 
> Which is puzzling, because both curl_7.18.0.bb and curl_7.52.1.bb depend on it.
> 
> Confused...

Actually, `curl` depend on `gnutls`, but `curl-native` does not. So not quite as confused, now.
It seems quite intricate under the hood.

> The newer jpeg sources have changes to the Visual Studio Express 2010 project files. All the jpeg source code is the same.

So that's a win !

**Comment by ralph-irving** [comment]
My guess would be that the gnutls dependency was originaly missed because it was already installed in the base OS, which it is on my system and like the gettext issue with git.

I still haven't found a reason why the tzdata package is built for my radio firmware either.

**Comment by mw9** [comment]
> I just tried to rebuild git with gettext-native added and it failed to download the gnutls-2.4.2.tar.bz2 sources.

Were you building `git` or `git-native` ?

There's a difference in dependencies:
`git` depends on `curl`.
`git-native` depends on `curl-native`

And:
>Actually, curl depend on gnutls, but curl-native does not. So not quite as confused, now.

`bitbake -g git-native` and peruse `depends.dot`.

**Comment by ralph-irving** [comment]
I was building git, but you are right, it should be git-native as curl is not included in the firmware.
The depends.dot file created by bitbake -g is handy.

**Comment by mw9** [comment]
> Thanks for the new radio firmware, I'll give that a go and report back.

OK, it updated fine, and the new `jive_alsa` seems to behave as expected. I'll leave it running on a 22050 Hz stream to double check that out, and check that there are no other surprises. If I have any issues I'll report them on the Forum thread.

I have just noted commits baec5dc60037d2cc75c8c5e0716b9c2d51cdb8b8 and 28636c0578c615018b6d549ac0f74d36a5863c42  which resolve the _Checksum error for jpegsrc.v8b.tar.gz_ and _Build failure git v2.25.5_ issues that I noted above. These align with what I have locally, and work.

I'll raise a separate issue regarding the updated time zones, if I can see anything that might be of value.

So I think appropriate to close at this piont.

**Comment by ralph-irving** [comment]
Sorry, I forgot to reference this issue in those commits.

**Comment by mw9** [comment]
> Sorry, I forgot to reference this issue in those commits.

No problem, I've been away for a few days and just wanted to "put it away" first.

I'll put up a new issue for timezone data, I think I've found a potentially very clean solution.
